In [1]:
%pip install ipython-sql

Note: you may need to restart the kernel to use updated packages.


In [2]:
%reload_ext sql
import csv, sqlite3

In [3]:
con = sqlite3.connect("my_data1.db")
cur = con.cursor()

In [4]:
!pip install -q pandas==1.1.5

In [5]:
%sql sqlite:///my_data1.db

In [6]:
import pandas as pd
df = pd.read_csv('/Users/noah/Desktop/Player Efficiency Analysis/CSV/hurts_dataset2.csv')

In [7]:
df.to_sql("HURTSTBL", con, if_exists='replace', index=False,method="multi")

46

In [8]:
sql select * from HURTSTBL

 * sqlite:///my_data1.db
Done.


[(13.0, 2020.0, '2020-12-13', 14.0, 'PHI', 'Home', 'NOR', 'W 24-21', 'Start', 30.0, 56.67, 167.0, 1.0, 0.0, 0.0, 18.0, 106.0, 0.0, 69.0, 1),
 (14.0, 2020.0, '2020-12-20', 15.0, 'PHI', 'Away', 'ARI', 'L 26-33', 'Start', 44.0, 54.55, 338.0, 3.0, 0.0, 6.0, 11.0, 63.0, 1.0, 83.0, 0),
 (15.0, 2020.0, '2020-12-27', 16.0, 'PHI', 'Away', 'DAL', 'L 17-37', 'Start', 39.0, 53.85, 342.0, 1.0, 2.0, 3.0, 9.0, 69.0, 0.0, 73.0, 0),
 (16.0, 2020.0, '2021-01-03', 17.0, 'PHI', 'Home', 'WAS', 'L 14-20', 'Start', 20.0, 35.0, 72.0, 0.0, 1.0, 1.0, 8.0, 34.0, 2.0, 50.0, 0),
 (17.0, 2021.0, '2021-09-12', 1.0, 'PHI', 'Away', 'ATL', 'W 32-6', 'Start', 35.0, 77.14, 264.0, 3.0, 0.0, 1.0, 7.0, 62.0, 0.0, 71.0, 1),
 (18.0, 2021.0, '2021-09-19', 2.0, 'PHI', 'Home', 'SFO', 'L 11-17', 'Start', 23.0, 52.17, 190.0, 0.0, 0.0, 2.0, 10.0, 82.0, 1.0, 58.0, 0),
 (19.0, 2021.0, '2021-09-27', 3.0, 'PHI', 'Away', 'DAL', 'L 21-41', 'Start', 39.0, 64.1, 326.0, 2.0, 2.0, 2.0, 9.0, 35.0, 0.0, 58.0, 0),
 (20.0, 2021.0, '2021-10-03', 4.0, 'PHI', 'Home', 'KAN', 'L 30-42', 'Start', 48.0, 66.67, 387.0, 2.0, 0.0, 3.0, 8.0, 47.0, 0.0, 74.0, 0),
 (21.0, 2021.0, '2021-10-10', 5.0, 'PHI', 'Away', 'CAR', 'W 21-18', 'Start', 37.0, 59.46, 198.0, 0.0, 1.0, 2.0, 9.0, 30.0, 2.0, 67.0, 1),
 (22.0, 2021.0, '2021-10-14', 6.0, 'PHI', 'Home', 'TAM', 'L 22-28', 'Start', 26.0, 46.15, 115.0, 1.0, 1.0, 2.0, 10.0, 44.0, 2.0, 52.0, 0),
 (23.0, 2021.0, '2021-10-24', 7.0, 'PHI', 'Away', 'LVR', 'L 22-33', 'Start', 34.0, 52.94, 236.0, 2.0, 0.0, 2.0, 13.0, 61.0, 0.0, 69.0, 0),
 (24.0, 2021.0, '2021-10-31', 8.0, 'PHI', 'Away', 'DET', 'W 44-6', 'Start', 14.0, 64.29, 103.0, 0.0, 0.0, 0.0, 7.0, 71.0, 0.0, 51.0, 1),
 (25.0, 2021.0, '2021-11-07', 9.0, 'PHI', 'Home', 'LAC', 'L 24-27', 'Start', 17.0, 64.71, 162.0, 1.0, 0.0, 1.0, 10.0, 62.0, 0.0, 57.0, 0),
 (26.0, 2021.0, '2021-11-14', 10.0, 'PHI', 'Away', 'DEN', 'W 30-13', 'Start', 23.0, 69.57, 178.0, 2.0, 1.0, 2.0, 13.0, 55.0, 0.0, 64.0, 1),
 (27.0, 2021.0, '2021-11-21', 11.0, 'PHI', 'Home', 'NOR', 'W 40-29', 'Start', 24.0, 54.17, 147.0, 0.0, 0.0, 3.0, 18.0, 69.0, 3.0, 78.0, 1),
 (28.0, 2021.0, '2021-11-28', 12.0, 'PHI', 'Away', 'NYG', 'L 7-13', 'Start', 31.0, 45.16, 129.0, 0.0, 3.0, 1.0, 8.0, 77.0, 0.0, 67.0, 0),
 (30.0, 2021.0, '2021-12-21', 15.0, 'PHI', 'Home', 'WAS', 'W 27-17', 'Start', 26.0, 76.92, 296.0, 1.0, 1.0, 3.0, 8.0, 38.0, 2.0, 71.0, 1),
 (31.0, 2021.0, '2021-12-26', 16.0, 'PHI', 'Home', 'NYG', 'W 34-10', 'Start', 29.0, 58.62, 199.0, 2.0, 0.0, 1.0, 2.0, 7.0, 0.0, 56.0, 1),
 (32.0, 2021.0, '2022-01-02', 17.0, 'PHI', 'Away', 'WAS', 'W 20-16', 'Start', 26.0, 65.38, 214.0, 0.0, 0.0, 1.0, 7.0, 44.0, 0.0, 61.0, 1),
 (34.0, 2022.0, '2022-09-11', 1.0, 'PHI', 'Away', 'DET', 'W 38-35', 'Start', 32.0, 56.25, 243.0, 0.0, 0.0, 1.0, 17.0, 90.0, 1.0, 77.0, 1),
 (35.0, 2022.0, '2022-09-19', 2.0, 'PHI', 'Home', 'MIN', 'W 24-7', 'Start', 31.0, 83.87, 333.0, 1.0, 1.0, 3.0, 11.0, 57.0, 2.0, 73.0, 1),
 (36.0, 2022.0, '2022-09-25', 3.0, 'PHI', 'Away', 'WAS', 'W 24-8', 'Start', 35.0, 62.86, 340.0, 3.0, 0.0, 3.0, 9.0, 20.0, 0.0, 70.0, 1),
 (37.0, 2022.0, '2022-10-02', 4.0, 'PHI', 'Home', 'JAX', 'W 29-21', 'Start', 25.0, 64.0, 204.0, 0.0, 1.0, 2.0, 16.0, 38.0, 1.0, 82.0, 1),
 (38.0, 2022.0, '2022-10-09', 5.0, 'PHI', 'Away', 'ARI', 'W 20-17', 'Start', 36.0, 72.22, 239.0, 0.0, 0.0, 2.0, 15.0, 61.0, 2.0, 73.0, 1),
 (39.0, 2022.0, '2022-10-16', 6.0, 'PHI', 'Home', 'DAL', 'W 26-17', 'Start', 25.0, 60.0, 155.0, 2.0, 0.0, 4.0, 9.0, 27.0, 0.0, 71.0, 1),
 (40.0, 2022.0, '2022-10-30', 8.0, 'PHI', 'Home', 'PIT', 'W 35-13', 'Start', 28.0, 67.86, 285.0, 4.0, 0.0, 3.0, 2.0, 10.0, 0.0, 46.0, 1),
 (41.0, 2022.0, '2022-11-03', 9.0, 'PHI', 'Away', 'HOU', 'W 29-17', 'Start', 27.0, 77.78, 243.0, 2.0, 0.0, 4.0, 9.0, 23.0, 0.0, 65.0, 1),
 (42.0, 2022.0, '2022-11-14', 10.0, 'PHI', 'Home', 'WAS', 'L 21-32', 'Start', 26.0, 65.38, 175.0, 2.0, 1.0, 1.0, 6.0, 28.0, 1.0, 50.0, 0),
 (43.0, 2022.0, '2022-11-20', 11.0, 'PHI', 'Away', 'IND', 'W 17-16', 'Start', 25.0, 72.0, 190.0, 1.0, 0.0, 3.0, 16.0, 86.0,

In [9]:
sql select distinct "Cmp%" from HURTSTBL order by "Cmp%" DESC

 * sqlite:///my_data1.db
Done.


[(83.87,),
 (78.26,),
 (77.78,),
 (77.14,),
 (76.92,),
 (76.32,),
 (74.36,),
 (74.19,),
 (73.91,),
 (72.22,),
 (72.0,),
 (69.57,),
 (67.86,),
 (67.74,),
 (67.57,),
 (66.67,),
 (65.79,),
 (65.38,),
 (64.71,),
 (64.29,),
 (64.1,),
 (64.0,),
 (63.64,),
 (62.86,),
 (62.22,),
 (62.16,),
 (60.0,),
 (59.46,),
 (58.62,),
 (58.06,),
 (57.78,),
 (57.14,),
 (56.67,),
 (56.25,),
 (54.55,),
 (54.17,),
 (53.85,),
 (52.94,),
 (52.17,),
 (46.15,),
 (45.16,),
 (35.0,)]

Here are all unique values for completion percentage that Jalen Hurts has recorded. Hurts's highest Completion Percentage was 83.87% while his lowest was 35%.

In [10]:
sql select distinct "Sacks" from HURTSTBL order by "Sacks" DESC

 * sqlite:///my_data1.db
Done.


[(6.0,), (5.0,), (4.0,), (3.0,), (2.0,), (1.0,), (0.0,)]

Here are all unique values for sacks in a game that Jalen Hurts has recorded. Hurts's highest number of sacks in a single game was 6 while his lowest was 0.

In [11]:
sql select distinct "Pass YDs" from HURTSTBL order by "Pass YDs" DESC

 * sqlite:///my_data1.db
Done.


[(387.0,),
 (380.0,),
 (342.0,),
 (340.0,),
 (338.0,),
 (333.0,),
 (326.0,),
 (319.0,),
 (315.0,),
 (303.0,),
 (298.0,),
 (296.0,),
 (285.0,),
 (280.0,),
 (279.0,),
 (277.0,),
 (264.0,),
 (243.0,),
 (239.0,),
 (236.0,),
 (229.0,),
 (217.0,),
 (214.0,),
 (207.0,),
 (204.0,),
 (200.0,),
 (199.0,),
 (198.0,),
 (193.0,),
 (190.0,),
 (178.0,),
 (175.0,),
 (170.0,),
 (167.0,),
 (162.0,),
 (155.0,),
 (153.0,),
 (150.0,),
 (147.0,),
 (129.0,),
 (115.0,),
 (103.0,),
 (72.0,)]

The highest recorded passing yards in a single game by Jalen Hurts is 387 while the lowest is 72. 

In [12]:
sql select distinct "Rush YDs" from HURTSTBL order by "Rush YDs" DESC

 * sqlite:///my_data1.db
Done.


[(157.0,),
 (106.0,),
 (90.0,),
 (86.0,),
 (82.0,),
 (77.0,),
 (72.0,),
 (71.0,),
 (69.0,),
 (65.0,),
 (63.0,),
 (62.0,),
 (61.0,),
 (57.0,),
 (55.0,),
 (47.0,),
 (44.0,),
 (38.0,),
 (37.0,),
 (36.0,),
 (35.0,),
 (34.0,),
 (30.0,),
 (29.0,),
 (28.0,),
 (27.0,),
 (23.0,),
 (21.0,),
 (20.0,),
 (13.0,),
 (12.0,),
 (10.0,),
 (7.0,),
 (6.0,)]

The highest recorded rushing yards in a single game by Jalen Hurts is 157 while the lowest is 6.

In [13]:
sql select "Opp", SUM("Pass YDs") as "Total Pass YDs" from HURTSTBL group by "Opp" order by SUM("Pass YDs") DESC

 * sqlite:///my_data1.db
Done.


[('WAS', 1735.0),
 ('DAL', 1030.0),
 ('NYG', 774.0),
 ('ARI', 577.0),
 ('KAN', 537.0),
 ('MIN', 526.0),
 ('SFO', 488.0),
 ('TAM', 392.0),
 ('TEN', 380.0),
 ('DET', 346.0),
 ('CHI', 315.0),
 ('NOR', 314.0),
 ('LAR', 303.0),
 ('PIT', 285.0),
 ('NYJ', 280.0),
 ('MIA', 279.0),
 ('ATL', 264.0),
 ('HOU', 243.0),
 ('LVR', 236.0),
 ('JAX', 204.0),
 ('BUF', 200.0),
 ('CAR', 198.0),
 ('IND', 190.0),
 ('DEN', 178.0),
 ('NWE', 170.0),
 ('LAC', 162.0),
 ('GNB', 153.0)]

This table shows the total number of passing yards by Opponent. Jalen has thrown the highest passing yards to Washington while he has thrown the lowest to Green Bay. The number of times each of these teams was played could be significantly biasing the numbers. 

In [14]:
sql select "Opp", ROUND(AVG("Pass YDs")) as "Avg Pass Yards" from HURTSTBL group by "Opp" order by AVG("Pass YDs") DESC

 * sqlite:///my_data1.db
Done.


[('TEN', 380.0),
 ('CHI', 315.0),
 ('LAR', 303.0),
 ('ARI', 289.0),
 ('PIT', 285.0),
 ('NYJ', 280.0),
 ('MIA', 279.0),
 ('KAN', 269.0),
 ('ATL', 264.0),
 ('MIN', 263.0),
 ('DAL', 258.0),
 ('WAS', 248.0),
 ('SFO', 244.0),
 ('HOU', 243.0),
 ('LVR', 236.0),
 ('JAX', 204.0),
 ('BUF', 200.0),
 ('CAR', 198.0),
 ('TAM', 196.0),
 ('NYG', 194.0),
 ('IND', 190.0),
 ('DEN', 178.0),
 ('DET', 173.0),
 ('NWE', 170.0),
 ('LAC', 162.0),
 ('NOR', 157.0),
 ('GNB', 153.0)]

Now we can see that Washington is closer to the middle and Jalen has probably played them more, skewing the total passing yards statisitic. Green Bay is still at the bottom and now Tennesse is at the top with 380 yards. 
NOTE: Both the top and bottom of this table are numbers from only one game played however. Not a long term average. 

In [27]:
sql select "Opp", ROUND(AVG("Cmp%"),2) as 'Comp %' from HURTSTBL group by "Opp" order by AVG("Cmp%") DESC

 * sqlite:///my_data1.db
Done.


[('MIN', 81.07),
 ('HOU', 77.78),
 ('ATL', 77.14),
 ('TEN', 74.36),
 ('MIA', 74.19),
 ('IND', 72.0),
 ('DEN', 69.57),
 ('PIT', 67.86),
 ('NWE', 66.67),
 ('LAR', 65.79),
 ('KAN', 65.16),
 ('LAC', 64.71),
 ('WAS', 64.2),
 ('JAX', 64.0),
 ('ARI', 63.39),
 ('DAL', 62.97),
 ('NYJ', 62.22),
 ('DET', 60.27),
 ('CHI', 59.46),
 ('CAR', 59.46),
 ('BUF', 58.06),
 ('NYG', 57.17),
 ('GNB', 57.14),
 ('NOR', 55.42),
 ('SFO', 54.98),
 ('TAM', 54.16),
 ('LVR', 52.94)]

This may be a better way of understanding Hurts's efficiency against a team. By looking at his completion percentage by team, we see how many of his passes were completed. From this table, we can gather that Hurts is most efficient against Minnesota and least efficient against Las Vegas. 

In [16]:
sql select "Opp", ROUND(AVG("Rush YDs")) as "Avg Rush Yards" from HURTSTBL group by "Opp" order by AVG("Rush YDs") DESC

 * sqlite:///my_data1.db
Done.


[('GNB', 157.0),
 ('NOR', 88.0),
 ('IND', 86.0),
 ('DET', 81.0),
 ('LAR', 72.0),
 ('BUF', 65.0),
 ('LAC', 62.0),
 ('ATL', 62.0),
 ('ARI', 62.0),
 ('LVR', 61.0),
 ('CHI', 61.0),
 ('DEN', 55.0),
 ('SFO', 51.0),
 ('NYJ', 47.0),
 ('MIN', 46.0),
 ('NYG', 44.0),
 ('DAL', 42.0),
 ('KAN', 38.0),
 ('JAX', 38.0),
 ('NWE', 37.0),
 ('TAM', 36.0),
 ('CAR', 30.0),
 ('WAS', 29.0),
 ('HOU', 23.0),
 ('MIA', 21.0),
 ('TEN', 12.0),
 ('PIT', 10.0)]

Due to the previous possible bias in the table for total passing yards, I decided not to do a total rushing yards table and skip to the averages. Hurts had the highest against Green Bay with 157 and the lowest against Pittsburgh with 10. 

In [17]:
sql select "Opp", ROUND(AVG("Int"),1) as "Avg Ints", SUM("Int") as "Total Ints" from HURTSTBL group by "Opp" order by AVG("Int") DESC

 * sqlite:///my_data1.db
Done.


[('NYJ', 3.0, 3.0),
 ('CHI', 2.0, 2.0),
 ('TAM', 1.5, 3.0),
 ('NYG', 1.0, 4.0),
 ('MIN', 1.0, 2.0),
 ('MIA', 1.0, 1.0),
 ('LAR', 1.0, 1.0),
 ('JAX', 1.0, 1.0),
 ('DEN', 1.0, 1.0),
 ('DAL', 1.0, 4.0),
 ('CAR', 1.0, 1.0),
 ('BUF', 1.0, 1.0),
 ('KAN', 0.5, 1.0),
 ('WAS', 0.4, 3.0),
 ('TEN', 0.0, 0.0),
 ('SFO', 0.0, 0.0),
 ('PIT', 0.0, 0.0),
 ('NWE', 0.0, 0.0),
 ('NOR', 0.0, 0.0),
 ('LVR', 0.0, 0.0),
 ('LAC', 0.0, 0.0),
 ('IND', 0.0, 0.0),
 ('HOU', 0.0, 0.0),
 ('GNB', 0.0, 0.0),
 ('DET', 0.0, 0.0),
 ('ATL', 0.0, 0.0),
 ('ARI', 0.0, 0.0)]

Hurts has thrown the most interceptions to the Giants but on average he has thrown the most per game to the Jets. 

In [18]:
sql select "Opp", ROUND(AVG("Sacks"),1) as "Avg Sacks", SUM("Sacks") as "Total Sacks" from HURTSTBL group by "Opp" order by AVG("Sacks") DESC

 * sqlite:///my_data1.db
Done.


[('KAN', 4.0, 8.0),
 ('HOU', 4.0, 4.0),
 ('ARI', 4.0, 8.0),
 ('MIN', 3.5, 7.0),
 ('PIT', 3.0, 3.0),
 ('NWE', 3.0, 3.0),
 ('MIA', 3.0, 3.0),
 ('IND', 3.0, 3.0),
 ('DAL', 3.0, 12.0),
 ('SFO', 2.5, 5.0),
 ('NYG', 2.3, 9.0),
 ('WAS', 2.0, 14.0),
 ('TEN', 2.0, 2.0),
 ('NYJ', 2.0, 2.0),
 ('LVR', 2.0, 2.0),
 ('JAX', 2.0, 2.0),
 ('GNB', 2.0, 2.0),
 ('DEN', 2.0, 2.0),
 ('CAR', 2.0, 2.0),
 ('BUF', 2.0, 2.0),
 ('TAM', 1.5, 3.0),
 ('NOR', 1.5, 3.0),
 ('LAR', 1.0, 1.0),
 ('LAC', 1.0, 1.0),
 ('CHI', 1.0, 1.0),
 ('ATL', 1.0, 1.0),
 ('DET', 0.5, 1.0)]

Hurts has been sacked the most by Washington but on average he has been sacked the most per game by Kansas City. Houston, and Arizona.

In [19]:
sql select "H/A", "Class" as "Result", COUNT("Class") as "Win/Loss Count" from HURTSTBL group by "H/A", "Class"

 * sqlite:///my_data1.db
Done.


[('Away', 0, 6), ('Away', 1, 17), ('Home', 0, 7), ('Home', 1, 16)]

This table shows Hurts's wins and losses based on whether the game was at Home or Away. He wins slightly more games Away but the numbers are almost identical. 

Something that stood out to me immediately from the above tables was that Green Bay appeared at the top of the rushing yards table and at the bottom of the passing yards table. Without looking into the tables beyond this, my hypothesis is the more rushing yards Hurts has the less passing yards he has or vice versa. Now I will find out if this is true. 

In [20]:
sql select ROUND(AVG("Pass YDs")) as "Over/Under AVG", Count("Pass YDs") as "Games Count" from HURTSTBL group by "Pass YDs" < (select AVG("Pass YDs") from HURTSTBL)

 * sqlite:///my_data1.db
Done.


[(302.0, 22), (172.0, 24)]

Out of the 46 games that we are using, 22 of those games he has thrown above his average in passing yards and 24 of those games he has thrown under his average in passing yards. 

In [21]:
sql select ROUND(AVG("Rush YDs")) as "Over/Under AVG", Count("Rush YDs") as "Games Count" from HURTSTBL where "Pass YDs" > (select AVG("Pass YDs") from HURTSTBL) group by "Rush YDs" < (select AVG("Rush YDs") from HURTSTBL)

 * sqlite:///my_data1.db
Done.


[(66.0, 9), (26.0, 13)]

Out of those 22 games where he threw over his average, 13 of them he has run under his average in rushing yards. This shows that when he is able to throw for more yards, he tends to run the ball less. Or when he can't run the ball very well in game, he tends to throw the ball more. 

In [22]:
sql select ROUND(AVG("Rush YDs")) as "Over/Under AVG", Count("Rush YDs") as "Games Count" from HURTSTBL where "Pass YDs" < (select AVG("Pass YDs") from HURTSTBL) group by "Rush YDs" < (select AVG("Rush YDs") from HURTSTBL)

 * sqlite:///my_data1.db
Done.


[(82.0, 11), (31.0, 13)]

When hurts passing yards are below his average, we can see that the hypothesis stays true. His rushing averages both went up. This means that the more passing yards he is getting, the less rushing yards he is inclined to produce and vice versa. But which is better? Which produces more wins? Rushing more and passing less or passing more and rushing less?

In [23]:
sql select "Class" as "Result", COUNT("Class") as "Total" from HURTSTBL where "Pass YDs" > (select AVG("Pass YDs") from HURTSTBL) and "Rush YDs" < (select AVG("Rush YDs") from HURTSTBL) group by "Class"

 * sqlite:///my_data1.db
Done.


[(0, 4), (1, 9)]

Wins and Loses for when Hurts has above average passing yards and below average rushing yards. Hurts wins 69.2% of the time in this scenario. 

In [24]:
sql select "Class" as "Result", COUNT("Class") as "Total" from HURTSTBL where "Pass YDs" < (select AVG("Pass YDs") from HURTSTBL) and "Rush YDs" > (select AVG("Rush YDs") from HURTSTBL) group by "Class"

 * sqlite:///my_data1.db
Done.


[(0, 3), (1, 8)]

Wins and Loses for when Hurts has below average passing yards and above average rushing yards. Hurts wins 72.7% of the time in this scenario. 

In [25]:
sql select "Class" as "Result", COUNT("Class") as "Total" from HURTSTBL where "Pass YDs" > (select AVG("Pass YDs") from HURTSTBL) and "Rush YDs" > (select AVG("Rush YDs") from HURTSTBL) group by "Class"

 * sqlite:///my_data1.db
Done.


[(0, 3), (1, 6)]

Wins and Loses for when Hurts has above average passing yards and above average rushing yards. Hurts wins 66.7% of the time in this scenario. 

In [26]:
sql select "Class" as "Result", COUNT("Class") as "Total" from HURTSTBL where "Pass YDs" < (select AVG("Pass YDs") from HURTSTBL) and "Rush YDs" < (select AVG("Rush YDs") from HURTSTBL) group by "Class"

 * sqlite:///my_data1.db
Done.


[(0, 3), (1, 10)]

Wins and Loses for when Hurts has below average passing yards and below average rushing yards. Hurts wins 76.9% of the time in this scenario. 

From these findings, it can be concluded that less passing yards seems to be the most productive play style for Jalen Hurts based off of the end goal of winning games. Rushing yards has some say in the matter in that when he rushes less he tends to win more. Hurts loses most frequently when he is both passing for above average yards and rushing for above average yards.